File: nucmed-modern/lib/schemas/article.ts
import { z } from "zod";
import { CATEGORIES } from "@/lib/config/constants";

export { CATEGORIES };

/**
 * Validation schemas for Article API
 */

export const ArticleStatusEnum = z.enum([
  "INGESTED",
  "PROCESSING",
  "DRAFT",
  "SCHEDULED",
  "PUBLISHED",
  "ARCHIVED",
  "FAILED",
]);

export type ArticleStatus = z.infer<typeof ArticleStatusEnum>;

/**
 * Schema for updating article (PATCH)
 */
export const ArticleUpdateSchema = z.object({
  title: z.string().min(1).max(500).optional(),
  content: z.string().min(1).optional(),
  excerpt: z.preprocess((val) => val === "" ? null : val, z.string().max(1000).optional().nullable()),
  // Relaxed validation to allow legacy categories during migration
  category: z.preprocess((val) => val === "" ? null : val, z.string().optional().nullable()),
  tags: z.array(z.string().min(1).max(50)).max(10).optional(),
  coverImageUrl: z.preprocess((val) => val === "" ? null : val, z.string().url().optional().nullable()),
  scheduledAt: z.preprocess((val) => val === "" ? null : val, z.coerce.date().optional().nullable()),
  publishedAt: z.preprocess((val) => val === "" ? null : val, z.coerce.date().optional().nullable()),
  originalPublishedAt: z.preprocess((val) => val === "" ? null : val, z.coerce.date().optional().nullable()),
  status: ArticleStatusEnum.optional(),
  significanceScore: z.preprocess((val) => val === "" ? null : val, z.coerce.number().int().min(1).max(10).optional().nullable()),
}).strict(); // Reject unknown fields

export type ArticleUpdate = z.infer<typeof ArticleUpdateSchema>;

/**
 * Schema for publishing article
 */
export const ArticlePublishSchema = z.object({
  scheduledAt: z.string().datetime().optional(), // If provided, schedule instead of immediate publish
});

export type ArticlePublish = z.infer<typeof ArticlePublishSchema>;

File: nuclear-news/apps/api/app/schemas/article.py
"""
Pydantic schemas for Article API
"""
from datetime import datetime
from uuid import UUID
from typing import Optional
from pydantic import BaseModel, Field


class ArticleBase(BaseModel):
    title: str = Field(..., max_length=500)
    excerpt: str
    content: str
    category: str = Field(..., max_length=100)
    tags: list[str] = []
    significance_score: Optional[int] = Field(None, ge=1, le=10)
    cover_image_url: Optional[str] = None


class ArticleCreate(ArticleBase):
    source_ids: list[UUID]
    original_urls: list[str]
    ai_model: Optional[str] = None


class ArticleUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=500)
    excerpt: Optional[str] = None
    content: Optional[str] = None
    category: Optional[str] = Field(None, max_length=100)
    tags: Optional[list[str]] = None
    cover_image_url: Optional[str] = None


class ArticleResponse(ArticleBase):
    id: UUID
    slug: str
    source_ids: list[UUID]
    original_urls: list[str]
    status: str
    published_at: Optional[datetime] = None
    ai_model: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ArticleList(BaseModel):
    items: list[ArticleResponse]
    total: int
    skip: int
    limit: int
